### PR TITLE
fix debezium reader thread lake

### LINKED ISSUE #1350: [mysql] The debezium reader resources exists thread leak risk
The debezium reader resources exists thread leak risk, we should release the thread timely

### PR DESCRIPTION
**Pull request by Matrix42**
this fix #1350

### CODE REVIEW

File: flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/debezium/reader/BinlogSplitReader.java [hunk 1]
```diff
@@ -174,6 +175,11 @@ public void close() {
             if (statefulTaskContext.getBinaryLogClient() != null) {
                 statefulTaskContext.getBinaryLogClient().disconnect();
             }
+            // set currentTaskRunning to false to terminate the
+            // while loop in MySqlStreamingChangeEventSource's execute method
+            currentTaskRunning = false;
+            executor.shutdown();
+            executor.awaitTermination(5, TimeUnit.SECONDS);
```
Comment (ruanhang1993) [inline flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/debezium/reader/BinlogSplitReader.java:182]: This timeout should be saved as a constant.

Comment (ruanhang1993) [inline flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/debezium/reader/BinlogSplitReader.java:182]: Maybe we could add a log when the `awaitTermination` method failed.

File: flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/debezium/reader/SnapshotSplitReader.java [hunk 2]
```diff
@@ -330,6 +331,8 @@ public void close() {
             if (statefulTaskContext.getBinaryLogClient() != null) {
                 statefulTaskContext.getBinaryLogClient().disconnect();
             }
+            executor.shutdown();
+            executor.awaitTermination(5, TimeUnit.SECONDS);
```
Comment (ruanhang1993) [inline flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/debezium/reader/SnapshotSplitReader.java:335]: Same commnets in the `BinlogSplitReader.java`.

File: flink-connector-mysql-cdc/src/test/java/com/ververica/cdc/connectors/mysql/debezium/reader/BinlogSplitReaderTest.java [hunk 3]
```diff
@@ -443,6 +443,7 @@ private List<String> readBinlogSplits(
                 }
             }
         }
+        snapshotSplitReader.close();
```
Comment (ruanhang1993) [inline flink-connector-mysql-cdc/src/test/java/com/ververica/cdc/connectors/mysql/debezium/reader/BinlogSplitReaderTest.java:446]: Do we need `assertExecutorIsTerminated(snapshotSplitReader)`?

**Review by ruanhang1993 — COMMENTED** [review]
Thanks for your work, @Matrix42 . I have left some comments.

**Comment by Matrix42** [comment]
Hi @ruanhang1993! Thanks for your review. I'v addressed the issue you mentioned, I'll rebase the master when the compile problem resolved.

**Comment by Matrix42** [comment]
Hi @ruanhang1993! Please take a look when you are available. The test failed is same as  #1144 

CC @leonardBang

File: flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/debezium/reader/SnapshotSplitReader.java [hunk 4]
```diff
@@ -328,6 +331,13 @@ public void close() {
             if (statefulTaskContext.getBinaryLogClient() != null) {
                 statefulTaskContext.getBinaryLogClient().disconnect();
             }
+            executor.shutdown();
+            boolean isShutdown =
+                    executor.awaitTermination(
+                            EXECUTOR_AWAIT_TERMINATION_TIMEOUT_SECONDS, TimeUnit.SECONDS);
+            if (!isShutdown) {
+                LOG.warn("The thread executor of SnapshotSplitReader wasn't shutdown properly.");
+            }
```
Comment (leonardBang) [inline flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/debezium/reader/SnapshotSplitReader.java:340]: Please check not NULL before call it.
```suggestion
       if (executorService != null) {
                executorService.shutdown();
                if (!executorService.awaitTermination(
                        READER_CLOSE_TIMEOUT, TimeUnit.MILLISECONDS)) {
                    LOG.warn(
                            "Failed to close the snapshot split reader in {} ms.",
                            READER_CLOSE_TIMEOUT);
                }
            }
```

File: flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/debezium/reader/SnapshotSplitReader.java [hunk 5]
```diff
@@ -86,6 +87,8 @@ public class SnapshotSplitReader implements DebeziumReader<SourceRecord, MySqlSp
     public AtomicBoolean hasNextElement;
     public AtomicBoolean reachEnd;
 
+    private static final long EXECUTOR_AWAIT_TERMINATION_TIMEOUT_SECONDS = 5;
```
Comment (leonardBang) [inline flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/debezium/reader/SnapshotSplitReader.java:90]:
```suggestion
    private static final long READER_CLOSE_TIMEOUT = 30L;
```
And let's rename `executor` to `executorService` for more precise meaning.

File: flink-connector-mysql-cdc/src/test/java/com/ververica/cdc/connectors/mysql/source/MySqlSourceTestBase.java [hunk 6]
```diff
@@ -86,4 +89,11 @@ public static void assertEqualsInOrder(List<String> expected, List<String> actua
         assertEquals(expected.size(), actual.size());
         assertArrayEquals(expected.toArray(new String[0]), actual.toArray(new String[0]));
     }
+
+    public static void assertExecutorIsTerminated(DebeziumReader<?, ?> reader) throws Exception {
+        Field executorField = reader.getClass().getDeclaredField("executor");
```
Comment (leonardBang) [inline flink-connector-mysql-cdc/src/test/java/com/ververica/cdc/connectors/mysql/source/MySqlSourceTestBase.java:94]: Use reflection is not cool here, if you really want to check `executorService` member of `SnapshotReader`, you can expose a method in `SnapshotReader` like:

```java
    @VisibleForTesting
    public ExecutorService getExecutorService() {
        return executorService;
    }
```

File: flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/debezium/reader/BinlogSplitReader.java [hunk 7]
```diff
@@ -182,12 +183,13 @@ public void close() {
...
-            boolean isShutdown =
-                    executor.awaitTermination(
-                            EXECUTOR_AWAIT_TERMINATION_TIMEOUT_SECONDS, TimeUnit.SECONDS);
-            if (!isShutdown) {
-                LOG.warn("The thread executor of BinlogSplitReader wasn't shutdown properly.");
+            if (executorService != null) {
+                executorService.shutdown();
+                if (executorService.awaitTermination(READER_CLOSE_TIMEOUT, TimeUnit.SECONDS)) {
+                    LOG.warn(
+                            "Failed to close the binlog split reader in {} ms.",
```
Comment (leonardBang) [inline flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/debezium/reader/BinlogSplitReader.java:190]:
```suggestion
                            "Failed to close the binlog split reader in {} seconds.",
```

File: flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/debezium/reader/SnapshotSplitReader.java [hunk 8]
```diff
@@ -331,18 +332,24 @@ public void close() {
...
-            boolean isShutdown =
-                    executor.awaitTermination(
-                            EXECUTOR_AWAIT_TERMINATION_TIMEOUT_SECONDS, TimeUnit.SECONDS);
-            if (!isShutdown) {
-                LOG.warn("The thread executor of SnapshotSplitReader wasn't shutdown properly.");
+            if (executorService != null) {
+                executorService.shutdown();
+                if (executorService.awaitTermination(READER_CLOSE_TIMEOUT, TimeUnit.SECONDS)) {
+                    LOG.warn(
+                            "Failed to close the snapshot split reader in {} ms.",
```
Comment (leonardBang) [inline flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/debezium/reader/SnapshotSplitReader.java:339]:
```suggestion
                            "Failed to close the snapshot split reader in {} seconds.",
```

File: flink-connector-mysql-cdc/src/test/java/com/ververica/cdc/connectors/mysql/debezium/reader/SnapshotSplitReaderTest.java [hunk 9]
```diff
@@ -413,7 +414,8 @@ private List<String> readTableSnapshotSplits(
         }
         snapshotSplitReader.close();
 
-        assertExecutorIsTerminated(snapshotSplitReader);
+        assertNull(snapshotSplitReader.getExecutorService());
+        assertTrue(snapshotSplitReader.getExecutorService().isTerminated());
```
Comment (leonardBang) [inline flink-connector-mysql-cdc/src/test/java/com/ververica/cdc/connectors/mysql/debezium/reader/SnapshotSplitReaderTest.java:418]: ? assertNotNull

File: flink-connector-mysql-cdc/src/test/java/com/ververica/cdc/connectors/mysql/debezium/reader/BinlogSplitReaderTest.java [hunk 10]
```diff
@@ -488,7 +491,8 @@ private List<String> readBinlogSplits(
         }
         binlogReader.close();
 
-        assertExecutorIsTerminated(binlogReader);
+        assertNull(snapshotSplitReader.getExecutorService());
+        assertTrue(snapshotSplitReader.getExecutorService().isTerminated());
```
Comment (leonardBang) [inline flink-connector-mysql-cdc/src/test/java/com/ververica/cdc/connectors/mysql/debezium/reader/BinlogSplitReaderTest.java:495]: ? assertNotNull

Comment (Matrix42) [reply flink-connector-mysql-cdc/src/test/java/com/ververica/cdc/connectors/mysql/debezium/reader/BinlogSplitReaderTest.java]: yes, I uesd wrong method.

**Review by leonardBang — APPROVED** [review]
Thanks @Matrix42 for the timely update, LGTM, let's wait the CI pass.

**Comment by Matrix42** [comment]
Thanks @leonardBang  for the review!

**Pull request merged**